File: apps/go/internal/chapter09/boston.go
package chapter09

import (
	"fmt"
	"strconv"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
)

const (
	// BostonTarget はボストンの住宅価格の正解の列。
	BostonTarget = "PRICE"
	// BostonCategory はカテゴリ値の列。
	BostonCategory = "CRIME"
)

// Scores は訓練データとテストデータの決定係数。
type Scores struct {
	Train float64
	Test  float64
}

// PrepareBoston は CRIME をダミー変数にし、分割してから訓練データの平均値で両方の欠損値を補完する。
func PrepareBoston(
	csvFile string,
	testSize float64,
	seed int64,
) (chapter02.TrainTestSplit[chapter02.Features, float64], error) {
	empty := chapter02.TrainTestSplit[chapter02.Features, float64]{}

	table, err := chapter02.LoadTable(csvFile)
	if err != nil {
		return empty, err
	}

	crimes := make([]string, 0, len(table.Rows))

	for _, row := range table.Rows {
		crime, err := row.Text(BostonCategory)
		if err != nil {
			return empty, err
		}

		crimes = append(crimes, crime)
	}

	encoded, err := Encode(table, BostonCategory, Categories(crimes))
	if err != nil {
		return empty, err
	}

	columns, rows, labels, err := chapter02.SplitFeaturesAndTarget(encoded, BostonTarget)
	if err != nil {
		return empty, err
	}

	prices := make([]float64, len(labels))

	for i, label := range labels {
		price, err := strconv.ParseFloat(label, 64)
		if err != nil {
			return empty, fmt.Errorf("%s を数値として読めません: %w", BostonTarget, err)
		}

		prices[i] = price
	}

	split, err := chapter02.SplitTrainTest(rows, prices, testSize, seed)
	if err != nil {
		return empty, err
	}

	means, err := chapter02.ColumnMeans(split.XTrain, columns)
	if err != nil {
		return empty, err
	}

	xTrain, err := chapter02.FillMissing(split.XTrain, columns, means)
	if err != nil {
		return empty, err
	}

	xTest, err := chapter02.FillMissing(split.XTest, columns, means)
	if err != nil {
		return empty, err
	}

	return chapter02.TrainTestSplit[chapter02.Features, float64]{
		XTrain: xTrain, XTest: xTest, TTrain: split.TTrain, TTest: split.TTest,
	}, nil
}

// ScoreFeatureSet は列から多項式特徴量を作って terms の項を選び、
// 訓練データで標準化してから線形回帰で学習し、決定係数を求める。
func ScoreFeatureSet(
	split chapter02.TrainTestSplit[chapter02.Features, float64],
	columns []string,
	terms []string,
) (Scores, error) {
	train, err := selectTerms(split.XTrain, columns, terms)
	if err != nil {
		return Scores{}, err
	}

	test, err := selectTerms(split.XTest, columns, terms)
	if err != nil {
		return Scores{}, err
	}

	standardizer, err := Fit(train)
	if err != nil {
		return Scores{}, err
	}

	xTrain, err := toRows(standardizer, train)
	if err != nil {
		return Scores{}, err
	}

	xTest, err := toRows(standardizer, test)
	if err != nil {
		return Scores{}, err
	}

	model, err := FitLinearModel(xTrain, split.TTrain)
	if err != nil {
		return Scores{}, err
	}

	return Scores{
		Train: RSquared(split.TTrain, model.Predict(xTrain)),
		Test:  RSquared(split.TTest, model.Predict(xTest)),
	}, nil
}

func selectTerms(x []chapter02.Features, columns, terms []string) ([]chapter02.Features, error) {
	expanded, err := Expand(x, columns)
	if err != nil {
		return nil, err
	}

	return Select(expanded, terms)
}

func toRows(standardizer Standardizer, x []chapter02.Features) ([][]float64, error) {
	transformed, err := standardizer.Transform(x)
	if err != nil {
		return nil, err
	}

	rows := make([][]float64, len(transformed))
	for i, features := range transformed {
		rows[i] = features.Values
	}

	return rows, nil
}

File: apps/ruby/lib/getting_started_ml/chapter09/boston.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 9 章の Boston データの前処理と、特徴量の組ごとの決定係数。
  module Chapter09
    # 正解の列。
    BOSTON_TARGET = "PRICE"
    # ダミー変数にするカテゴリの列。
    BOSTON_CATEGORY = "CRIME"

    # 訓練データとテストデータの決定係数。
    Scores = Data.define(:train, :test)

    module_function

    # CRIME をダミー変数にしてから、第 2 章の関数で「特徴量と正解に分ける → シード付きで分割する →
    # 訓練データの平均値で両方の欠損値を補完する」を行う。
    def prepare_boston(csv_file, test_size:, seed:)
      columns, rows, prices = boston_features_and_prices(Chapter02::Table.load(csv_file))
      split = Chapter02.split_train_test(rows, prices, test_size:, seed:)
      means = Chapter02.column_means(split.x_train, columns)

      split.with(x_train: Chapter02.fill_missing(split.x_train, columns, means),
                 x_test: Chapter02.fill_missing(split.x_test, columns, means))
    end

    # CRIME をダミー変数にしてから、特徴量の列・行・価格に分ける。価格は数値に読み直す。
    def boston_features_and_prices(table)
      crimes = table.rows.map { |row| row.text(BOSTON_CATEGORY) }
      encoded = encode(table, BOSTON_CATEGORY, categories(crimes))
      columns, rows, labels = Chapter02.split_features_and_target(encoded, BOSTON_TARGET)

      [columns, rows, labels.map { |label| Float(label) }]
    end

    # 多項式特徴量を作って項を選び、訓練データで標準化の平均と標準偏差を求めてから学習し、決定係数を測る。
    def score_feature_set(split, columns, terms)
      x_train, x_test = standardized_terms(split, columns, terms)
      model = LinearModel.fit(x_train, split.t_train)

      Scores.new(train: r_squared(split.t_train, model.predict(x_train)),
                 test: r_squared(split.t_test, model.predict(x_test)))
    end

    # 訓練データとテストデータで項を作って選び、訓練データの平均と標準偏差で標準化した値の並びにする。
    def standardized_terms(split, columns, terms)
      train, test = [split.x_train, split.x_test].map { |x| select(expand(x, columns), terms) }
      standardizer = Standardizer.fit(train)

      [train, test].map { |x| standardizer.transform(x).map(&:values) }
    end
  end
end
